File: destack-ts/src/language/core/builtin/struct.ts
import type { Encoding, PackedCache, StructDefinition } from "@destack/language/core";
import { ObjectKind, StructType } from "@destack/language/core/builtin/common";
import { ACTIVE_SESSION, ENCODERS } from "@destack/language/core/builtin/const";
import { BuiltinObject, BuiltinObjectClass } from "@destack/language/core/builtin/object";
import type { Session } from "@destack/language/core/runtime/session";
import type { StructTypeMapping } from "@destack/language/mapping";
import { registerStructClass } from "@destack/language/registry";
import { BinaryWriter } from "@destack/language/core/runtime/binary";

/** A Struct is an ordered collection of Properties. */
export abstract class Struct extends BuiltinObject {
  static readonly metatype: StructType;
  static readonly __kind__: ObjectKind = ObjectKind.STRUCT;
  static readonly __isStruct__: boolean = true;
  static readonly __definition__: StructDefinition;

  constructor(_session: Session | null) {
    _session = _session ?? ACTIVE_SESSION.get();
    super(_session);
  }

  get metatype(): StructType {
    return (this.constructor as typeof Struct).metatype;
  }
}
registerStructClass(StructType.STRUCT, Struct);

/** A frozen Struct is a Struct that is immutable. */
export abstract class StructFrozen extends Struct {
  static readonly __isFrozen__: boolean = true;

  /** Cached hash of the Struct. */
  readonly _hash: number | null = null;
  /** Cached repr of the Struct. */
  readonly _repr: string | null = null;
  /** Cached packed representations (first N = each Encoding, next N = each Encoding as bytes). */
  readonly _packedCache: PackedCache[] | null = null;

  _invalidateFrozenCache(): void {
    // frozen Structs should be immutable, but sometimes we need to break out of that
    // @ts-ignore
    this._hash = null;
    // @ts-ignore
    this._repr = null;
    // @ts-ignore
    this._packedCache = null;
  }

  override pack(encoding: Encoding): any {
    // check if we have a cached packed representation
    if (this._packedCache != null) {
      for (const cached of this._packedCache) {
        if (cached.encoding === encoding && !cached.isBytes) {
          return cached.packed;
        }
      }
    }
    // pack the object
    const encoder = ENCODERS[encoding]!;
    const packedObject = encoder.packObject(ObjectKind.STRUCT, this.metatype, this);
    // cache the result
    const newCache: PackedCache = {
      encoding: encoding,
      isBytes: false,
      packed: packedObject,
    };
    if (this._packedCache == null) {
      // @ts-expect-error(readonly)
      this._packedCache = [newCache];
    } else {
      // @ts-expect-error(readonly)
      this._packedCache = [...this._packedCache, newCache];
    }
    return packedObject;
  }

  packBinary(encoding: Encoding): Uint8Array {
    // check if we have a cached packed bytes representation
    if (this._packedCache != null) {
      for (const cached of this._packedCache) {
        if (cached.encoding === encoding && cached.isBytes) {
          return cached.packed;
        }
      }
    }
    // pack the object as bytes
    const encoder = ENCODERS[encoding]!;
    const writer = new BinaryWriter();
    encoder.packObjectBinary(ObjectKind.STRUCT, this.metatype, this, writer);
    const packedObjectBytes = writer.toBytes();
    // cache the result
    const newCache: PackedCache = {
      encoding: encoding,
      isBytes: true,
      packed: packedObjectBytes,
    };
    if (this._packedCache == null) {
      // @ts-expect-error(readonly)
      this._packedCache = [newCache];
    } else {
      // @ts-expect-error(readonly)
      this._packedCache = [...this._packedCache, newCache];
    }
    return packedObjectBytes;
  }
}

/** A Struct class. */
type StructConstructor<S extends Struct = Struct> = new (...args: any[]) => S;
type AbstractStructConstructor<S extends Struct = Struct> = abstract new (...args: any[]) => S;
export type StructClass<S extends Struct = Struct> = (
  | StructConstructor<S>
  | AbstractStructConstructor<S>
) &
  BuiltinObjectClass<any> & {
    metatype: StructType;
    __definition__: StructDefinition;
  };

/** Check if a value is a Struct of a specific type. */
export function isStruct<T extends StructType>(
  value: any,
  structType?: T,
): value is StructTypeMapping[T] {
  return value instanceof Struct && (structType === undefined || value.metatype === structType);
}

File: destack-py/destack/language/core/builtin/struct.py
from typing import (
    TYPE_CHECKING,
    Any,
    ClassVar,
    Self,
    cast,
    dataclass_transform,
)

from destack.language.registry import STRUCT_CLASS_BY_TYPE, STRUCT_TYPE_BY_CLASS
from destack.utils.env import IS_DEV, IS_TEST
from destack.utils.log import get_logger
from destack.utils.telemetry import get_tracer

from .builtin import EnumType, ObjectKind, ObjectStability, StructType
from .declaration import StructDeclaration, TagDeclaration, builtin_method
from .object import Object, _process_object_cls
from .property import _PROPERTY_SPECIFIERS, builtin_property_runtime

if TYPE_CHECKING:
    from destack.language import StructDefinition

# pyright: reportIncompatibleVariableOverride=false

logger = get_logger(__name__)
tracer = get_tracer(__name__)
type_ = type


def _process_struct_cls(
    cls: type["Struct"],
    struct_type: StructType,
    stability: ObjectStability,
    is_frozen: bool,
    is_abstract: bool,
    is_final: bool,
    tags: tuple["TagDeclaration", ...],
) -> type["Struct"]:
    # bases
    inherits: list[StructType] = []
    all_enum_types: list[EnumType] = []
    if cls.__name__ != "Struct" and cls.__name__ != "StructFrozen":
        for base in cls.__mro__:
            if issubclass(base, Struct):
                if base.metatype not in inherits:
                    inherits.append(base.metatype)
                for enum_type in base.__declaration__.self_enum_types:
                    if enum_type not in all_enum_types:
                        all_enum_types.append(enum_type)

    # declaration
    declaration = StructDeclaration(
        # meta
        cls=cls,
        type=struct_type,
        id=struct_type.value,
        kind=ObjectKind.STRUCT,
        stability=stability,
        is_abstract=is_abstract,
        is_frozen=is_frozen,
        is_final=is_final,
        # inherits
        base_type=inherits[0] if inherits else None,
        inherits=list(reversed(inherits)),
        inherited_by=[],
        extended_by=[],
        # content
        properties=[],
        methods=[],
        constants=[],
        tags=list(tags),
        # associations
        enum_types=list(all_enum_types),
        self_enum_types=list(all_enum_types),
    )

    # process object class
    cls, _ = _process_object_cls(cast(type["Struct"], cls), declaration)
    cls.metatype = struct_type

    # register struct
    if cls.__name__ != "StructFrozen":
        assert cls.__name__ == "Struct" or issubclass(cls, Struct), (
            f"struct class {cls} is not a Struct"
        )
        if struct_type in STRUCT_CLASS_BY_TYPE:
            raise ValueError(
                f"struct class conflict for {struct_type}: {cls}, {STRUCT_CLASS_BY_TYPE[struct_type]}"
            )
        STRUCT_CLASS_BY_TYPE[struct_type] = cls
        STRUCT_TYPE_BY_CLASS[cls] = struct_type

    # sanity check
    if IS_DEV or IS_TEST:
        # non-abstract structs must have properties
        if not is_abstract and not any(
            not prop.is_runtime_only for prop in cls.__declaration__.properties
        ):
            raise ValueError(f"{cls.__name__} is not abstract but has no properties")
        # abstract objects cannot extend non-abstract objects
        if is_abstract and cls.__bases__ and not cls.__bases__[0].__is_abstract__:
            raise ValueError(
                f"{cls.__name__} is abstract but extends non-abstract {cls.__bases__[0].__name__}"
            )
        # cannot be both abstract and final
        if is_abstract and is_final:
            raise ValueError(f"{cls.__name__} cannot be both abstract and final")
        # final classes must be annotated with @final
        if is_final != getattr(cls, "__final__", False):
            raise ValueError(
                f"{cls.__name__} has @final={getattr(cls, '__final__', False)} but is_final={is_final}"
            )
        # final objects cannot be extended
        if any(
            hasattr(base, "__declaration__") and base.__declaration__.is_final
            for base in cls.__bases__
        ):
            bad_base = next(
                base
                for base in cls.__bases__
                if hasattr(base, "__declaration__") and base.__declaration__.is_final
            )
            raise ValueError(f"{cls.__name__} extends final {bad_base.__name__}")

    return cast(type["Struct"], cls)


@dataclass_transform(kw_only_default=True, field_specifiers=_PROPERTY_SPECIFIERS)
def builtin_struct(
    struct_type: StructType,
    *,
    frozen: bool = False,
    is_abstract: bool = False,
    is_final: bool = False,
    stability: ObjectStability = ObjectStability.DYNAMIC,
    tags: tuple["TagDeclaration", ...] = (),
):
    """Register a class as a concrete struct for the given struct type."""

    def decorate(cls: type) -> type:
        cls = _process_struct_cls(
            cls=cast(type["Struct"], cls),
            struct_type=struct_type,
            stability=stability,
            is_frozen=frozen,
            is_abstract=is_abstract,
            is_final=is_final,
            tags=tags,
        )
        return cls

    return decorate


@builtin_struct(StructType.STRUCT, is_abstract=True)
class Struct(Object):
    """A Struct is an ordered collection of Properties."""

    # meta
    metatype: ClassVar[StructType]
    metakind: ClassVar[ObjectKind] = ObjectKind.STRUCT
    __declaration__: ClassVar["StructDeclaration"]
    __definition__: ClassVar["StructDefinition"]

    def __eq__(self, other: Any):
        """Equals the Struct contents."""
        raise NotImplementedError  # generated


@builtin_struct(
    StructType.STRUCT,
    frozen=True,  # type: ignore (frozen can't inherit from non-frozen usually, but it's fine for us)
    is_abstract=True,
)
class StructFrozen(Struct):
    """An immutable Struct."""

    """Cached hash of the Struct."""
    _hash: "int | None" = builtin_property_runtime()
    """Cached repr of the Struct."""
    _repr: "str | None" = builtin_property_runtime()

    def _invalidate_frozen_cache(self) -> None:
        # frozen Structs should be immutable, but sometimes we need to break out of that
        object.__setattr__(self, "_hash", None)
        object.__setattr__(self, "_repr", None)

    @builtin_method(60)
    def clone(self, **override: Any) -> Self:
        """Clone the Struct with new values."""
        kwargs: dict[str, Any] = {}
        for prop in self.__properties__.values():
            if not prop.is_runtime_only:
                kwargs[prop.name] = getattr(self, prop.name)
        kwargs.update(override)
        return self.__class__(**kwargs)
